libmscore/drumset.cpp

```diff
@@ -19,33 +19,59 @@ namespace Ms {
 Drumset* smDrumset;           // standard midi drumset
 Drumset* gpDrumset;           // guitar pro drumset
 
+//---------------------------------------------------------
+//   Drumset
+//---------------------------------------------------------
+
+Drumset::Drumset()
+      {
+      _drum.resize(DRUM_INSTRUMENTS);
+      for (int i = 0; i < DRUM_INSTRUMENTS; ++i) {
+            _drum[i].notehead = NoteHead::Group::HEAD_INVALID;
+            _drum[i].line     = 0;
+            _drum[i].shortcut = 0;
+            _drum[i].voice    = 0;
+            _drum[i].stemDirection = MScore::Direction::UP;
+            _drum[i].pitch = i;
+            }
+      }
+
+std::vector<DrumInstrument> Drumset::drumsByIndex() const
+      {
+      auto d = _drum;
+      std::sort(d.begin(), d.end());
+      return d;
+      }
+
 //---------------------------------------------------------
 //   save
 //---------------------------------------------------------
 
 void Drumset::save(Xml& xml) const
       {
-      for (int i = 0; i < 128; ++i) {
-            if (!isValid(i))
+      auto d = drumsByIndex();
+      for (int n = 0; n < DRUM_INSTRUMENTS; ++n) {
+            DrumInstrument di = d[n];
```

Comment:
const DrumInstrument& is more effective. In your case you need to copy data from `d[n]` to `di`. If data of the 'di' is not changed, reference is good.